src/common/Throttle.h

```diff
@@ -330,4 +331,77 @@ class OrderedThrottle {
   uint32_t waiters = 0;
 };
 
+
+class TokenBucketThrottle {
+
+  struct Bucket {
+    CephContext *cct;
+    const std::string name;
+    std::atomic<uint64_t> remain = { 0 }, max = { 0 };
+
+    Bucket(CephContext *cct, const std::string& n, uint64_t m)
+      : cct(cct), name(n),
+	remain(m), max(m)
+    {
+    }
+
+    uint64_t get(uint64_t c);
+    uint64_t put(uint64_t c);
+    void set_max(uint64_t m);
+  };
+
+  struct Blocker {
+    uint64_t tokens_requested;
+    Context *ctx;
+
+    Blocker(uint64_t _tokens_requested, Context* _ctx)
+      : tokens_requested(_tokens_requested), ctx(_ctx) {}
+  };
+
+  CephContext *m_cct;
+  Bucket m_throttle;
+  uint64_t m_avg = 0;
+  SafeTimer *m_timer;
+  Mutex *m_timer_lock;
+  FunctionContext *m_token_ctx = nullptr;
+  Mutex m_blockers_lock;
+  list<Blocker> m_blockers;
+  Mutex m_lock;
+
+public:
+TokenBucketThrottle(CephContext *cct, uint64_t capacity, uint64_t avg,
+		    SafeTimer *timer, Mutex *timer_lock);
+
+~TokenBucketThrottle();
+
+template <typename T, typename I, void(T::*MF)(int, I*)>
+bool get(uint64_t c, T *handler, I *item) {
+  if (0 == m_throttle.max)
+    return false;
+
+  bool waited = false;
+
+  Mutex::Locker lock(m_lock);
+  uint64_t got = m_throttle.get(c);
+  if (got < c) {
+    // Got not enough tokens, add a blocker for it.
+    Context *ctx = new FunctionContext([this, handler, item](int r) {
+	(handler->*MF)(r, item);
+      });
+    m_blockers.emplace_back(c - got, ctx);
```

Comment:
Not holding new ```m_blockers_lock``` -- I suspect holding the lock here will re-introduce the lockdep issue. Might need to use the alternative implementation of having the ```add_tokens``` prepare a list of blockers to complete (so that you don't drop and re-acquire two locks in a loop), set a ```m_callback_in_progress``` flag, drop the locks, invoke the callbacks, re-acquire the lock(s), and signal the condition variable.

Reply:
Yes, this should be protected by m_blockers_lock. and it might be lockdep because I will make the m_lock only protect the ```throttle.get()```, then the m_lock and m_blockers_lock will be not dependent as what it is in add_tokens().

In addition, actually, when I seperate the m_blockers_lock from m_lock, I am not dropping and re-acquire the locks in loop now. Then we don't need a m_callback_in_progress to avoid race between ~TokenBucketThrottle() and add_tokens(), the m_timer_lock can do it.

Reply:
Oops, wait, this would be a cycle lock. let me think more about it.